_posts/2022-9-5-Diplomat @ Google.md

```diff
@@ -0,0 +1,503 @@
+different declarations:
+```rust
+#[diplomat::bridge]
+mod ffi {
+    #[diplomat::opaque]
+    pub struct Opaque {
+        // ...
+    }
+
+    pub struct Input<'i> {
+        data: &'i Opaque,
+    }
+
+    pub struct Output<'o> {
+        data: &'o Opaque,
+    }
+
+    impl<'a> Input<'a> {
+        fn get_data(input: Input<'a>) -> Output<'a> {
+            let out = Output { data: input.data }
+            out
+        }
+    }
+}
+```
+The lifetimes in the declarations of `Input` and `Output` have completely
+different names. We can't just track `'a` all the way down, which sucks because
+that would have been easy. Instead, we have to track lifetimes by the index they
+appear at in generic arguments, since that's what rustc does.
+
+## The Big Picture
+Let's put this all together and walk through an example of a method with
+lifetime bounds, non-opaque structs in the input and output positions,
+and renamed lifetimes:
+```rust
+#[diplomat::bridge]
+mod ffi {
+    #[diplomat::opaque]
+    pub struct Opaque {
+        // ...
+    }
+
+    pub struct Input<'i> {
+        data: &'i Opaque,
+    }
+
+    pub struct Output<'o> {
+        data: &'o Opaque,
+    }
+
+    impl<'a> Input<'a> {
+        fn get_data<'b>(input: Input<'a>) -> Output<'b>
+        where
+            'a: 'b,
+        {
+            let out = Output { data: input.data }
+            out
+        }
+    }
+}
+```
+The first thing that Diplomat does when it encounters `Input::get_data` is
+examine the return type signature, `Output<'b>`. Since it's a non-opaque struct,
+we look in the declaration of `Output` and store a mapping
+saying that the position of `'o` in the context of `Output`'s declaration
+corresponds with the position of `'b` in `Output<'b>` in the method. Next, we
+see the `data: &'o Opaque` field, and use the mapping to determine that it's
+actually `&'b Opaque` in the context of the method. We then remark that
+`out.data` relies on `'b`, see that there's no more fields to explore in the
+field traversal, and finish.
+
+Once we know which lifetimes to look out for (just `'b`), we traverse the types
+of the
+inputs, which consists of just `Input<'a>`. Since `Input` is a non-opaque struct,
+we look in the declaration of `Input` and store a mapping saying that the
+position of `'i` in the context of `Input`'s declaration corresponds with the
+position of `'a` in `Input<'a>` in the method. Inside of `Input`s declaration
+we see the `data: &'i Opaque` field, so we map the lifetime `'i` back to `'a`,
+and then perform a depth-first search in the lifetime graph to find all shorter
+lifetimes.  This search results in finding `'b`. After confirming that `'b`
+appears in the output, we remark that `input.data` is one of the input fields
+that lives at least `'b`.
+
+In code generation, we put these pieces of information together to determine
+that `out.data` must be outlived by `input.data`, which is exactly what we
+were aiming for.
```

Comment:
suggestion: probably worth mentioning what the final result is from a codegen POV. i.e. mention that when `get_data` is codegenned in JS, it includes some code that does something like `out.data.owner.push(input.data)` or something.